utils/land-patterns.stanza

```diff
@@ -24,6 +26,60 @@ public defn ref-label () :
   ref-label(0.0, 0.0)
 
 ; ====== Land pattern utilities ===============================================
+; Apply a shape to the soldermask layer of a pcb-pad. 
+public defn apply-soldermask (mask-shape:Shape):
+  inside pcb-pad:
+    layer(SolderMask(Top)) = mask-shape
+    if pad-type(self) == TH:
+      layer(SolderMask(Bottom)) = mask-shape
+
+; Create the soldermask layers based on the pad shape, using a fixed factor.
+; when expand? is true the soldermask shape is computed with expand(), else it is
+; computed with offset()
+public defn apply-soldermask (amount:Double, expand?:True|False):
+  val func = expand when expand? else offset
+  inside pcb-landpattern:
+    apply-soldermask(func(pad-shape(self), amount))
+
+; Apply a default soldermask layer to a pcb-pad, using the SOLDERMASK-EXPANSION
+; design variable and offset() method.
+public defn apply-soldermask ():
+  apply-soldermask(SOLDERMASK-EXPANSION, false)
+
+; A "solder mask defined pad" is one where the soldermask opening is smaller than 
+; the pad shape. This method takes a pad as an argument and converts it into a 
+; soldermask-defined version based on its soldermask layer, if it exists, or the
+; pad shape if it does not.
+public defn soldermask-defined (p:Pad):
+  ; extract the pad data before conversion
+  val mask?   = find({specifier(_) is SolderMask}, layers(p))  
+  ; special case: the mask is smaller than the shape already
+  if (mask? is-not False) and 
+     (min-width([pad-shape(p)]) > min-width([shape $ (mask? as LayerShape)])) :
+    p
+  else :
+    pcb-pad SMD-p:
+      name  = to-string("%_, SMD" % [name(p)])
+      type  = pad-type(p)
+      shape = pad-shape(p)
```

Comment:
I am confused. Is it the same shape that is used on line 67?

Reply:
No. This is a `pcb-pad` definition.

Reply:
is this line needed?

Reply:
Yes, all pads require a shape.